flow/directives.py

```diff
@@ -0,0 +1,357 @@
+        if not isinstance(directive, _DirectivesItem):
+            raise TypeError(
+                f"Expected a _DirectivesItem object. Received {type(directive)}.")
+        elif directive._name in self._directive_definitions:
+            raise ValueError(f"Cannot define {directive._name} twice.")
+        else:
+            self._directive_definitions[directive._name] = directive
+            self._defined_directives[directive._name] = directive._default
+
+    def _set_defined_directive(self, key, value):
+        try:
+            self._defined_directives[key] = self._directive_definitions[key](value)
+        except Exception as err:
+            raise DirectivesError(f'Error setting directive "{key}"') from err
+
+    def __getitem__(self, key):
+        try:
+            value = self._defined_directives[key]
+            return self._directive_definitions[key]._finalize(value, self)
+        except KeyError:
+            try:
+                return self._user_directives[key]
+            except KeyError:
+                raise KeyError(f"{key} not in directives.")
+
+    def __setitem__(self, key, value):
+        if key in self._directive_definitions:
+            self._set_defined_directive(key, value)
+        else:
+            self._user_directives[key] = value
+
+    def __delitem__(self, key):
+        if key in self._directive_definitions:
+            self._defined_directives[key] = self._directive_definitions[key]._default
+        else:
+            del self._user_directives[key]
+
+    def __iter__(self):
+        yield from self._defined_directives
+        yield from self._user_directives
+
+    def __len__(self):
+        return len(self._defined_directives) + len(self._user_directives)
+
+    def __str__(self):
+        return str(dict(self))
+
+    def __repr__(self):
+        return f"_Directives({str(self)})"
+
+    def update(self, other, aggregate=False, jobs=None, parallel=False):
+        if aggregate:
+            self._aggregate(other, jobs=jobs, parallel=parallel)
+        else:
+            super().update(other)
+
+    def evaluate(self, jobs):
+        for key, value in self.items():
+            self[key] = _evaluate(value, jobs)
+
+    def _aggregate(self, other, jobs=None, parallel=False):
+        self.evaluate(jobs)
+        other.evaluate(jobs)
+        agg_func_attr = "_parallel" if parallel else "_serial"
+        for name in self._defined_directives:
+            agg_func = getattr(self._directive_definitions[name], agg_func_attr)
+            default_value = self._directive_definitions[name]._default
+            other_directive = other.get(name, default_value)
+            directive = self[name]
+            if other_directive is None:
+                continue
+            else:
+                self._defined_directives[name] = agg_func(directive,
+                                                          other_directive)
+
+
+def _evaluate(value, jobs):
```

Comment:
This can be a static function in the `_Directives` class

Reply:
We can move it into a helper function of `_aggregate`. `jobs` could be `None` when combining directives without a job such as `directives.update(other_directives, aggregate=True)`. In practice this may not be used often, but it makes sense to allow.